lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -299,15 +300,188 @@ fn connect_rdp_inner(
+                        &mut CGOSharedDirectoryReadRequest {
+                            completion_id: req.completion_id,
+                            directory_id: req.directory_id,
+                            path: c_string.as_ptr(),
+                            path_length: req.path.len() as u32,
+                            offset: req.offset,
+                            length: req.length,
+                        },
+                    );
+
+                    if err != CGOErrCode::ErrCodeSuccess {
+                        return Err(RdpError::TryError(String::from(
+                            "call to tdp_sd_read_request failed",
+                        )));
+                    }
+                }
+                Ok(())
+            }
+            Err(_) => {
+                return Err(RdpError::TryError(format!(
+                    "path contained characters that couldn't be converted to a C string: {}",
+                    req.path
+                )));
+            }
+        }
+    });
+
+    let tdp_sd_write_request = Box::new(move |req: SharedDirectoryWriteRequest| -> RdpResult<()> {
+        debug!("sending: {:?}", req);
+        match CString::new(req.path.clone()) {
+            Ok(c_string) => {
+                unsafe {
+                    let err = tdp_sd_write_request(
+                        go_ref,
+                        &mut CGOSharedDirectoryWriteRequest {
+                            completion_id: req.completion_id,
+                            directory_id: req.directory_id,
+                            offset: req.offset,
+                            path: c_string.as_ptr(),
+                            path_length: req.path.len() as u32,
+                            write_data_length: req.write_data.len() as u32,
+                            write_data: req.write_data.as_ptr() as *mut u8,
+                        },
+                    );
+
+                    if err != CGOErrCode::ErrCodeSuccess {
+                        return Err(RdpError::TryError(String::from(
+                            "call to tdp_sd_write_failed",
+                        )));
+                    }
+                }
+                Ok(())
+            }
+            Err(_) => {
+                return Err(RdpError::TryError(format!(
+                    "path contained characters that couldn't be converted to a C string: {}",
+                    req.path
+                )));
+            }
+        }
+    });
+
     // Client for the "rdpdr" channel - smartcard emulation and drive redirection.
-    let rdpdr = rdpdr::Client::new(
-        params.cert_der,
-        params.key_der,
+    let rdpdr = rdpdr::Client::new(rdpdr::Config {
+        cert_der: params.cert_der,
+        key_der: params.key_der,
         pin,
-        params.allow_directory_sharing,
+        allow_directory_sharing: params.allow_directory_sharing,
         tdp_sd_acknowledge,
         tdp_sd_info_request,
-    );
+        tdp_sd_create_request,
```

Comment:
I'm not suggesting a change now, but wanted to point out that this API seems fragile. If you mistakenly get the order of these callbacks wrong everything will break and the compiler won't be able to help you.